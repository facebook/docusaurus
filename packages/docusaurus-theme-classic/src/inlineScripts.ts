/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import type {SiteStorage} from '@docusaurus/types';
import type {ThemeConfig} from '@docusaurus/theme-common';

// Support for ?docusaurus-theme=dark
const ThemeQueryStringKey = 'docusaurus-theme';

// Support for ?docusaurus-data-mode=embed&docusaurus-data-myAttr=42
const DataQueryStringPrefixKey = 'docusaurus-data-';

export function getThemeInlineScript({
  colorMode: {defaultMode, respectPrefersColorScheme},
  siteStorage,
}: {
  colorMode: ThemeConfig['colorMode'];
  siteStorage: SiteStorage;
}): string {
  // Need to be inlined to prevent dark mode FOUC
  // Make sure the key is the same as the one in the color mode React context
  // Currently defined in: `docusaurus-theme-common/src/contexts/colorMode.tsx`
  const themeStorageKey = `theme${siteStorage.namespace}`;

  /* language=js */
  return `(function() {
    var defaultMode = '${defaultMode}';
    var respectPrefersColorScheme = ${respectPrefersColorScheme};

    function setDataThemeAttribute(theme) {
      document.documentElement.setAttribute('data-theme', theme);
    }

    function getQueryStringTheme() {
      try {
        return new URLSearchParams(window.location.search).get('${ThemeQueryStringKey}')
      } catch (e) {
      }
    }

    function getStoredTheme() {
      try {
        return window['${siteStorage.type}'].getItem('${themeStorageKey}');
      } catch (err) {
      }
    }

    var initialTheme = getQueryStringTheme() || getStoredTheme();
    if (initialTheme !== null) {
      setDataThemeAttribute(initialTheme);
    } else {
      if (
        respectPrefersColorScheme &&
        window.matchMedia('(prefers-color-scheme: dark)').matches
      ) {
        setDataThemeAttribute('dark');
      } else if (
        respectPrefersColorScheme &&
        window.matchMedia('(prefers-color-scheme: light)').matches
      ) {
        setDataThemeAttribute('light');
      } else {
        setDataThemeAttribute(defaultMode === 'dark' ? 'dark' : 'light');
      }
    }
  })();`;
}

/* language=js */
export const DataAttributeQueryStringInlineJavaScript = `
(function() {
  try {
    const entries = new URLSearchParams(window.location.search).entries();
    for (var [searchKey, value] of entries) {
      if (searchKey.startsWith('${DataQueryStringPrefixKey}')) {
        var key = searchKey.replace('${DataQueryStringPrefixKey}',"data-")
        document.documentElement.setAttribute(key, value);
      }
    }
  } catch(e) {}
})();
`;

// We always render the announcement bar html on the server, to prevent layout
// shifts on React hydration. The theme can use CSS + the data attribute to hide
// the announcement bar asap (before React hydration)
export function getAnnouncementBarInlineScript({
  siteStorage,
}: {
  siteStorage: SiteStorage;
}): string {
  // Duplicated constant. Unfortunately we can't import it from theme-common, as
  // we need to support older nodejs versions without ESM support
  // TODO: import from theme-common once we support Node.js with ESM support
  // + move all those announcementBar stuff there too
  const AnnouncementBarDismissStorageKey = `docusaurus.announcement.dismiss${siteStorage.namespace}`;
  const AnnouncementBarDismissDataAttribute =
    'data-announcement-bar-initially-dismissed';

  /* language=js */
  return `(function() {
  function isDismissed() {
    try {
      return localStorage.getItem('${AnnouncementBarDismissStorageKey}') === 'true';
    } catch (err) {}
    return false;
  }
  document.documentElement.setAttribute('${AnnouncementBarDismissDataAttribute}', isDismissed());
})();`;
}
