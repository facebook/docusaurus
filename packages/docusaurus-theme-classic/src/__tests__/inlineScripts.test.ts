/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {getThemeInlineScript} from '../inlineScripts';

describe('inlineScripts', () => {
  it('inline javascript for default options', () => {
    expect(
      getThemeInlineScript({
        colorMode: {
          disableSwitch: false,
          defaultMode: 'light',
          respectPrefersColorScheme: false,
        },
        siteStorage: {
          type: 'localStorage',
          namespace: '',
        },
      }),
    ).toMatchSnapshot();
  });

  it('inline javascript for prefers color scheme and no switch', () => {
    expect(
      getThemeInlineScript({
        colorMode: {
          disableSwitch: true,
          defaultMode: 'light',
          respectPrefersColorScheme: true,
        },
        siteStorage: {
          type: 'localStorage',
          namespace: '',
        },
      }),
    ).toMatchSnapshot();
  });
});
