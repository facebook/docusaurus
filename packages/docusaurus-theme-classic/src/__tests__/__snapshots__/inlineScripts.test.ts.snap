// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`inlineScripts inline javascript for default options 1`] = `
"(function() {
  function getSystemColorMode() {
    return window.matchMedia('(prefers-color-scheme: dark)').matches ? 'dark' : 'light';
  }
  function getQueryStringTheme() {
    try {
      return new URLSearchParams(window.location.search).get('docusaurus-theme')
    } catch (e) {}
  }
  function getStoredTheme() {
    try {
      return window['localStorage'].getItem('theme');
    } catch (err) {}
  }
  var initialTheme = getQueryStringTheme() || getStoredTheme();
  document.documentElement.setAttribute('data-theme', initialTheme || 'light');
  document.documentElement.setAttribute('data-theme-choice', initialTheme || 'light');
})();"
`;

exports[`inlineScripts inline javascript for prefers color scheme and no switch 1`] = `
"(function() {
  function getSystemColorMode() {
    return window.matchMedia('(prefers-color-scheme: dark)').matches ? 'dark' : 'light';
  }
  var initialTheme;
  document.documentElement.setAttribute('data-theme', initialTheme || getSystemColorMode());
  document.documentElement.setAttribute('data-theme-choice', initialTheme || 'system');
})();"
`;
